[bug] Compatible cached packages not resolved correctly

<!--
  Please don't forget to update the issue title.
  Include all applicable information to help us reproduce your problem.

  To help us debug your issue please explain:
-->

### Environment Details (include every applicable attribute)
  * Operating System+version: Windows 10 64
  * Compiler+version: VS 2019 16
  * Conan version: 1.41.0
  * Python version: 3.8.8
  * Revisions enabled: True

### Steps to reproduce (Include if Applicable)

* Assume packages A, B, C, D and two build settings Settings1 and Settings2 resulting in different package ids.
* B requires A
* C requires A 
* D requires both B and C
* A, B and C have been built with Settings1 and each override package_id to provide a compatible package for Settings2.
* D needs to be built with Settings2

If you call `conan install` on D with Settings2 we observer the following:
* A, B and C are reported as missing but a compatible package has been found: 
`A/1.0@user/channel: Main binary package 'f278b6f5a018dfdaf3f1aaf111c91b35b8dbeec5' missing. Using compatible package 'f85619e3f4a4d53fb5b48d3c40437fc2f08c69e6'`
`B/1.0@user/channel: Main binary package 'f278b6f5a018dfdaf3f1aaf111c91b35b8dbeec5' missing. Using compatible package 'f85619e3f4a4d53fb5b48d3c40437fc2f08c69e6'`
`C/1.0@user/channel: Main binary package 'f278b6f5a018dfdaf3f1aaf111c91b35b8dbeec5' missing. Using compatible package 'f85619e3f4a4d53fb5b48d3c40437fc2f08c69e6'`
* A might end up multiple times within the self.info.requires dict used in the package_id methods of B and C due to different package ids which leads to wrong hashes being created. 
* If we workaround the previous issue by throwing out the multiple occurrances of the same package requires within package_id, the wrong package binary is finally used for A:
`  File "c:\programdata\anaconda3\lib\site-packages\conans\client\installer.py", line 560, in _handle_node_cache`
`    assert os.path.isdir(package_folder), ("Package '%s' folder must exist: %s\n"`
`AssertionError: Package 'A/1.0@user/channel:f278b6f5a018dfdaf3f1aaf111c91b35b8dbeec5' folder must exist: C:\.conan\3785e5\1`

Perhaps this might be related to #9446

### Possible Fix ?

* From debugging we found that there exist two nodes in the conan graph for package A, which originally point to the same package id. 
* For the first node A `_evalute_node` from file `graph_binaries.py` is called and compatible packages are computed. Then, the _package_id of that node is set to the compatible package id `f85619...` and the node.binary is set to cached mode.
* For the second node A `_evaluate_is_cached` in the same file encounters the first node A (`previous_nodes`) and sets its binary also cached as well. However, the _package_id is not modified.

If we set the _package_id to the previous_node everything works fine but we are not sure if this has any unwanted side effects:
`node._package_id = previous_node._package_id`

I hope that this helps in resolving the bug.

<!--
  Your log content should be related to the bug description, it can be:
  - Conan command output
  - Server output (Artifactory, conan_server)
-->

I guess the reason for multiple entries in the graph is that one package declares a requirement as private, while another package uses it as normal dependency.

It can be reproduced with 3 packages, where pkga is used as private requirement by pkgb and both pkga and pkgb are requirements of pkgc.

I've changed your code to the following test case:

```
def test_compatible_diamond():
    # https://github.com/conan-io/conan/issues/9880
    client = TestClient()
    conanfile = '''
from conans import ConanFile
class Pkg(ConanFile):
    settings = "build_type"

    def package_id(self):
        compatible_pkg = self.info.clone()
        if self.settings.build_type == "Debug":
            compatible_pkg.settings.build_type = "Release"
            self.compatible_packages.append(compatible_pkg)

    {}
    '''

    client.save({"pkga/conanfile.py": conanfile.format(""),
                 "pkgb/conanfile.py": conanfile.format('''
    def requirements(self):
        self.requires("pkga/0.1", "private")'''),
                 "pkgc/conanfile.py": conanfile.format('''
    def requirements(self):
        self.requires("pkga/0.1")
        self.requires("pkgb/0.1")''')
                 })
    client.run("create pkga pkga/0.1@ -s build_type=Release")
    client.run("create pkgb pkgb/0.1@ -s build_type=Release")
    client.run("create pkgc pkgc/0.1@ -s build_type=Debug")
    print(client.out)
```

But it turns out that this bug a rather tricky one, as its occurrence is non-deterministic and I wasn't able to reproduce it from the pytest call, at all. However, if I make the call to create pkgc from the command line, it will fail roughly 3 out of 10 times.

I ran this with python 3.8 and conan 1.41.0 .

This is the output in case it fails:

```
C:\Projects\conan\pkgid-debug>conan create pkgc pkgc/0.1@ -s build_type=Debug
Exporting package recipe
pkgc/0.1: The stored package has not changed
pkgc/0.1: Using the exported files summary hash as the recipe revision: 6fe7e1268fe335f082c6fa3b3271b4f2
pkgc/0.1: Exported revision: 6fe7e1268fe335f082c6fa3b3271b4f2
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Debug
compiler=Visual Studio
compiler.runtime=MDd
compiler.version=16
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

pkga/0.1: Main binary package '5a67a79dbc25fd0fa149a0eb7a20715189a0d988' missing. Using compatible package '4024617540c4f240a6a5e8911b0de9ef38a11a72'
pkgb/0.1: Main binary package '7a8b95543b4e3738be01d064adc3f6c0449c9df5' missing. Using compatible package '96a5c7bd97e984b4bd30026de5908eea95ec48bd'
pkgc/0.1: Forced build from source
Installing package: pkgc/0.1
Requirements
    pkga/0.1 from local cache - Cache
    pkga/0.1 from local cache - Cache
    pkgb/0.1 from local cache - Cache
    pkgc/0.1 from local cache - Cache
Packages
    pkga/0.1:4024617540c4f240a6a5e8911b0de9ef38a11a72 - Skip
    pkga/0.1:5a67a79dbc25fd0fa149a0eb7a20715189a0d988 - Cache
    pkgb/0.1:96a5c7bd97e984b4bd30026de5908eea95ec48bd - Cache
    pkgc/0.1:3c5ae3b9400202e3edbf274032f2ca1f3b8c14c5 - Build

Installing (downloading, building) binaries...
pkga/0.1: Already installed!
Traceback (most recent call last):
  File "c:\programdata\anaconda3\lib\site-packages\conans\client\command.py", line 2232, in run
    method(args[0][1:])
  File "c:\programdata\anaconda3\lib\site-packages\conans\client\command.py", line 382, in create
    info = self._conan.create(args.path, name=name, version=version, user=user,
  File "c:\programdata\anaconda3\lib\site-packages\conans\client\conan_api.py", line 93, in wrapper
    return f(api, *args, **kwargs)
  File "c:\programdata\anaconda3\lib\site-packages\conans\client\conan_api.py", line 388, in create
    create(self.app, ref, graph_info, remotes, update, build_modes,
  File "c:\programdata\anaconda3\lib\site-packages\conans\client\cmd\create.py", line 84, in create
    deps_install(app=app,
  File "c:\programdata\anaconda3\lib\site-packages\conans\client\manager.py", line 83, in deps_install
    installer.install(deps_graph, remotes, build_modes, update, profile_host, profile_build,
  File "c:\programdata\anaconda3\lib\site-packages\conans\client\installer.py", line 316, in install
    self._build(nodes_by_level, keep_build, root_node, profile_host, profile_build,
  File "c:\programdata\anaconda3\lib\site-packages\conans\client\installer.py", line 451, in _build
    self._handle_node_cache(node, keep_build, processed_package_refs, remotes)
  File "c:\programdata\anaconda3\lib\site-packages\conans\client\installer.py", line 560, in _handle_node_cache
    assert os.path.isdir(package_folder), ("Package '%s' folder must exist: %s\n"
AssertionError: Package 'pkga/0.1:5a67a79dbc25fd0fa149a0eb7a20715189a0d988' folder must exist: C:\conan\data\pkga\0.1\_\_\package\5a67a79dbc25fd0fa149a0eb7a20715189a0d988
```

As you can see, pkga is listed twice as requirement and the correct package id is chosen initially, however while resolving the packages from cache, it skips the correct one.

Hope this helps.

--- a/conans/client/graph/graph_binaries.py
+++ b/conans/client/graph/graph_binaries.py
@@ -157,6 +157,10 @@ class GraphBinariesAnalyzer(object):
                 node.binary = previous_node.binary
             node.binary_remote = previous_node.binary_remote
             node.prev = previous_node.prev
+
+            # this line fixed the compatible_packages with private case.
+            # https://github.com/conan-io/conan/issues/9880
+            node._package_id = previous_node.package_id
             return True
         self._evaluated[pref] = [node]
 

--- a/conans/test/integration/package_id/compatible_test.py
+++ b/conans/test/integration/package_id/compatible_test.py
@@ -523,6 +523,35 @@ class CompatibleIDsTest(unittest.TestCase):
         self.assertIn("pkg/0.1@user/stable:cb054d0b3e1ca595dc66bc2339d40f1f8f04ab31", client.out)
         self.assertIn("pkg/0.1@user/stable: Already installed!", client.out)
 
+    def test_compatible_diamond(self):
+        # https://github.com/conan-io/conan/issues/9880
+        client = TestClient()
+        conanfile = textwrap.dedent("""
+            from conan import ConanFile
+            class Pkg(ConanFile):
+                {}
+                settings = "build_type"
+                def package_id(self):
+                    if self.settings.build_type == "Debug":
+                        compatible_pkg = self.info.clone()
+                        compatible_pkg.settings.build_type = "Release"
+                        self.compatible_packages.append(compatible_pkg)
+                """)
+
+        client.save({"pkga/conanfile.py": conanfile.format(""),
+                     "pkgb/conanfile.py": conanfile.format('requires = ("pkga/0.1", "private"), '),
+                     "pkgc/conanfile.py": conanfile.format('requires = "pkga/0.1"'),
+                     "pkgd/conanfile.py": conanfile.format('requires = "pkgb/0.1", "pkgc/0.1"')
+                     })
+        client.run("create pkga pkga/0.1@ -s build_type=Release")
+        client.run("create pkgb pkgb/0.1@ -s build_type=Release")
+        client.run("create pkgc pkgc/0.1@ -s build_type=Release")
+
+        client.run("install pkgd -s build_type=Debug")
+        assert "pkga/0.1: Main binary package '5a67a79dbc25fd0fa149a0eb7a20715189a0d988' missing" \
+               in client.out
+        assert "pkga/0.1:4024617540c4f240a6a5e8911b0de9ef38a11a72 - Cache" in client.out
+
 
 def test_msvc_visual_incompatible():
     conanfile = GenConanfile().with_settings("os", "compiler", "build_type", "arch")

